hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockReconstructedStripeInputStream.java

```diff
@@ -0,0 +1,421 @@
+    ByteBuffer buf = ByteBuffer.allocate(repConfig.getEcChunkSize());
+    return buf;
+  }
+
+  private void flipInputs() {
+    for (ByteBuffer b : decoderInputBuffers) {
+      if (b != null) {
+        b.flip();
+      }
+    }
+  }
+
+  private void clearParityBuffers() {
+    for (int i = getRepConfig().getData();
+         i < getRepConfig().getRequiredNodes(); i++) {
+      if (decoderInputBuffers[i] != null) {
+        decoderInputBuffers[i].clear();
+      }
+    }
+  }
+
+  protected void loadDataBuffersFromStream() throws IOException {
+    for (int i = 0; i < dataIndexes.size(); i++) {
+      BlockExtendedInputStream stream =
+          getOrOpenStream(i, dataIndexes.get(i));
+      ByteBuffer b = decoderInputBuffers[dataIndexes.get(i)];
+      while (b.hasRemaining()) {
+        int read = stream.read(b);
+        if (read == EOF) {
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Take the populated input buffers and missing indexes and create the
+   * outputs. Note that the input buffers have to be "ready for read", ie they
+   * need to have been flipped after their data was loaded. The created outputs
+   * are "ready to read" by the underlying decoder API, so there is no need to
+   * flip them after the call. The decoder reads all the inputs leaving the
+   * buffer position at the end, so the inputs are flipped after the decode so
+   * we have a complete set of "outputs" for the EC Stripe which are ready to
+   * read.
+   * @throws IOException
+   */
+  private void decodeStripe() throws IOException {
+    decoder.decode(decoderInputBuffers, missingIndexes, decoderOutputBuffers);
+    flipInputs();
+  }
+
+  @Override
+  public synchronized boolean hasSufficientLocations() {
+    // The number of locations needed is a function of the EC Chunk size. If the
+    // block length is <= the chunk size, we should only have one data location.
+    // If it is greater than the chunk size but less than chunk_size * 2, then
+    // we must have two locations. If it is greater than chunk_size * data_num,
+    // then we must have all data_num locations.
+    // The remaining data locations (for small block lengths) can be assumed to
+    // be all zeros.
+    // Then we need a total of dataNum blocks available across the available
+    // data, parity and padding blocks.
+    ECReplicationConfig repConfig = getRepConfig();
+    int expectedDataBlocks = calculateExpectedDataBlocks(repConfig);
+    int availableLocations =
+        availableDataLocations() + availableParityLocations();
+    int paddedLocations = repConfig.getData() - expectedDataBlocks;
+
+    if (availableLocations + paddedLocations >= repConfig.getData()) {
+      return true;
+    } else {
+      LOG.warn("There are insufficient locations. {} available {} padded {} " +
+          "expected", availableLocations, paddedLocations, expectedDataBlocks);
+      return false;
+    }
+  }
+
```

Comment:
Below method can moved to super class and use?
Same calculation is present in ECBlockInputStream#hasSufficientLocations

Reply:
Yea, you are correct - this method can go into the super class and be re-used in this class and the super class. I've made that change.